src/storage/src/render/upsert.rs

```diff
@@ -152,19 +153,29 @@ where
         None
     };
 
+    // Break `previous` into:
+    // On the one hand, "Ok" and "Err(UpsertError)", which we know how to deal with, and,
+    // On the other hand, "Err(everything eles)", which we don't.
+    let (previous, mut errs) = previous.ok_err(|(d, t, r)| match d {
+        Ok(row) => Ok((Ok(row), t, r)),
+        Err(DataflowError::EnvelopeError(EnvelopeError::Upsert(err))) => Ok((Err(err), t, r)),
+        Err(err) => Err((err, t, r)),
+    });
+
     let upsert_output = upsert_core(
         stream,
         predicates,
         position_or,
         as_of_frontier,
         upsert_envelope,
-        previous_ok,
+        previous.as_collection(),
         previous_token,
     );
-    let (mut oks, mut errs) = upsert_output.ok_err(|(data, time, diff)| match data {
+    let (mut oks, errs2) = upsert_output.ok_err(|(data, time, diff)| match data {
```

Comment:
`errs2` is not very descriptive, what kind of permanent errors can `upsert_core` add?

Reply:
`errs2` is used throughout the rest of this function.

`upsert_core` does not add any new permanent errors -- all the errors it emits into `errs2` should be retractable.